src/click/types.py

```diff
@@ -733,8 +733,8 @@ def convert(self, value, param, ctx):
         is_dash = self.file_okay and self.allow_dash and rv in (b"-", "-")
 
         if not is_dash:
-            if self.resolve_path:
-                rv = os.path.realpath(rv)
+            if self.resolve_path and os.path.islink(rv):
```

Comment:
`islink` and `readlink` needs to be a separate step within the `if` block before `realpath`.